src/section/section_chain.rs

```diff
@@ -0,0 +1,1114 @@
+    ///
+    /// When the trusted key is on the main branch, on the other hand:
+    ///
+    /// ```ascii-art
+    /// D->E->F
+    ///    |
+    ///    +->G
+    /// ```
+    ///
+    /// Then such chain is impossible to forge because the adversary would have to have access to
+    /// the secret key corresponding to `D` in order to validly sign `E`. Thus such chain can be
+    /// safely considered trusted.
+    pub fn check_trust<'a, I>(&self, trusted_keys: I) -> bool
+    where
+        I: IntoIterator<Item = &'a bls::PublicKey>,
+    {
+        let trusted_keys: HashSet<_> = trusted_keys.into_iter().collect();
+        let mut index = self.tree.len();
+
+        loop {
+            let (key, parent_index) = if index > 0 {
+                let block = &self.tree[index - 1];
+                (&block.key, Some(block.parent_index))
+            } else {
+                (&self.root, None)
+            };
+
+            if trusted_keys.contains(key) {
+                return true;
+            }
+
+            if let Some(next_index) = parent_index {
+                index = next_index;
+            } else {
+                return false;
+            }
+        }
+    }
+
+    /// Compare the two keys by their position in the chain. The key that is higher (closer to the
+    /// last key) is considered `Greater`. If exactly one of the keys is not in the chain, the other
+    /// one is implicitly considered `Greater`. If none are in the chain, they are considered
+    /// `Equal`.
+    pub fn cmp_by_position(&self, lhs: &bls::PublicKey, rhs: &bls::PublicKey) -> Ordering {
+        match (self.index_of(lhs), self.index_of(rhs)) {
+            (Some(lhs), Some(rhs)) => lhs.cmp(&rhs),
+            (Some(_), None) => Ordering::Greater,
+            (None, Some(_)) => Ordering::Less,
+            (None, None) => Ordering::Equal,
+        }
+    }
+
+    /// Returns the number of blocks in the chain. This is always >= 1.
+    pub fn len(&self) -> usize {
+        1 + self.tree.len()
+    }
+
+    fn insert_block(&mut self, new_block: Block) -> usize {
+        // Find the index into `self.tree` to insert the new key at. All the keys above will be
+        // pushed upwards.
+        let insert_at = self
+            .tree
+            .iter()
+            .enumerate()
+            .skip(new_block.parent_index)
+            .find(|(_, block)| {
+                block.parent_index != new_block.parent_index || block.key >= new_block.key
+            })
+            .map(|(index, _)| index)
+            .unwrap_or(self.tree.len());
+
+        // If the key already exists in the chain, do nothing but still return success to make the
+        // `insert` operation idempotent.
+        if self.tree.get(insert_at).map(|block| &block.key) != Some(&new_block.key) {
+            self.tree.insert(insert_at, new_block);
+
+            // Adjust the parent indices of the keys whose parents are above the inserted key.
+            for block in &mut self.tree {
```

Comment:
this update only need to be carried out for the blocks AFTER the inserted one AND whose parent is AFTER the inserted one as well ?

i.e. you can just iterate start from self.tree[insert_at + 2] ?

Reply:
Good point, but if the new block is the last block then `insert_at + 2` would be an invalid index. I'm going to use `insert_at + 1` instead which should always work.